Forward and backward binary compatibility between 2.10.0 and 2.10.1-RC1

adriaanm opened the pull request:
Start running mima both ways, filtering out failures using @gkossakowski's [contribution to mima](https://github.com/gkossakowski/migration-manager/tree/13-cli-filters), which should be in 0.1.5-SNAPSHOT soon, to make sure bc.run doesn't fail (even where it should).

Each commit in this PR comments out the filters it makes irrelevant, until we only need to filter out permitted binary incompatibilities.

We only allow binary incompatibilities in scala.reflect.internal.

Reverts backports of:
- SI-6521, overrides Range#head to be faster
- SI-4664 Make scala.util.Random Serializable
- SI-6548 reflection now correctly enters jinners

Also reverts/refactors:
- cosmetic renamings in runtime reflection
- duplicate tailImpl as a private method
- bring back SerializeStart from fa3b804
- can't add new class BatchingExecutor
- refactor AdaptedForkJoinTask, uncaughtExceptionHandler
    - SI-6989 privateWithin is now populated in reflect (refactored, fix preserved)

review by/ping @retronym, @xeno-by, @gkossakowski, @paulp, @jamesiry, @phaller, @viktorklang and the rest of the world

adriaanm committed:
[nomaster] run mima both ways, filter out failures

Using @gkossakowski's contribution to mima (in 0.1.5-SNAPSHOT),
make sure bc.run doesn't fail by filtering out all binary incompatibilities.

Each subsequent commit will comment out the filters it makes irrelevant,
until we only need to filter out permitted binary incompatibilities.

We only allow binary incompatibilities in scala.reflect.internal.

adriaanm committed:
[nomaster] Revert "Fixes SI-6521, overrides Range#head to be faster"

This reverts commit a557a973608a75c7a02f251bbcf49fe6f6b6655e.

This is necessary to maintain binary compatibility with 2.10.0.

Mima says:

    matchName="scala.collection.immutable.Range.head"
    problemName=IncompatibleResultTypeProblem

The bridge method appeared because result is now Int, whereas
the super-method's result type erases to Object

adriaanm committed:
[nomaster] Revert "SI-4664 Make scala.util.Random Serializable"

Also revert "SI-4664 [Make scala.util.Random Serializable] Add test case"

This reverts commit 0b92073a38f9d1823f051ac18173078bfcfafc8a.
This reverts commit 2aa66bec86fd464712b0d15251cc400ff9d52821.

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm committed:
[nomaster] duplicate tailImpl as a private method

Reworks d526f8bd74.
This is necessary to maintain binary compatibility with 2.10.0.

    matchName="scala.collection.mutable.MutableList.tailImpl"
    problemName=MissingMethodProblem

adriaanm committed:
[nomaster] bring back SerializeStart from fa3b8040eb

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm committed:
[nomaster] can't add new class BatchingExecutor

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm committed:
[nomaster] refactor AdaptedForkJoinTask, uncaughtExceptionHandler

Inlined AdaptedForkJoinTask, made uncaughtExceptionHandler private[this].

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm committed:
[nomaster] Revert "cosmetic renamings in runtime reflection"

This reverts commit 0429f0fd9224499cd8b606490d04b1a8dcffbca8.

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm committed:
[nomaster] Revert "SI-6548 reflection now correctly enters jinners"

This reverts commit 54a84a36d5b435a787d93ca48d45399136c7e162.

This is necessary to maintain binary compatibility with 2.10.0.

run/t6989.check had to be updated as it also (indirectly) tested SI-6548

Conflicts:
	test/files/lib/javac-artifacts.jar.desired.sha1
	test/files/run/t6548.check
	test/files/run/t6548/Test_2.scala

xeno-by commented:
Why not allow incompatibilities in scala.reflect.runtime? That's also an internal package.

retronym commented:
> Why not allow incompatibilities in scala.reflect.runtime? That's also an internal package.

:+1:

xeno-by commented:
Okay, we have two public definitions there: `scala.reflect.runtime.universe` and `scala.reflect.runtime.currentMirror`. But the rest is supposed to be internal. Everything except `JavaUniverse` there is private to either scala or reflect.

`JavaUniverse` is ugly, though. It's the return type of  `scala.reflect.runtime.universe`, so we can't make it private. Would be great to make it a public trait and hide the implementation class behind it. Though that'll probably be not compatible :(

update. Not it's not the return type of `ru`. See the subsequent comment from me after getting a sleep.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -675,11 +675,115 @@ object Future {
+    }
+
+    /** Override this to define which runnables will be batched. */
+    private def batchable(runnable: Runnable): Boolean = runnable match {
```
Fewer lines with the following I guess (Philipp removed Batchable which was a case below):

```
runnable.isInstanceOf[OnCompleteRunnable]
```

viktorklang replied:
and it's only called in 1 place, so inline the check

adriaanm replied:
ok

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -675,11 +675,115 @@ object Future {
+     * in the calling thread synchronously. It must enqueue/handoff the Runnable.
+     */
+    // invariant: if "_tasksLocal.get ne null" then we are inside BatchingRunnable.run; if it is null, we are outside
+    private val _tasksLocal = new ThreadLocal[List[Runnable]]()
```
InternalCallbackExecutor could extend ThreadLocal now to avoid the indirection

adriaanm replied:
I'll leave that for another PR

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -675,11 +675,115 @@ object Future {
-    override protected def unbatchedExecute(r: Runnable): Unit =
-      r.run()
+  private[concurrent] object InternalCallbackExecutor extends ExecutionContext with java.util.concurrent.Executor {
+    protected def unbatchedExecute(r: Runnable): Unit = r.run()
```
this does not need be protected and can be final now

adriaanm replied:
ok

viktorklang commented on src/library/scala/concurrent/impl/ExecutionContextImpl.scala:
```diff
@@ -25,7 +25,7 @@ private[scala] class ExecutionContextImpl private[impl] (es: Executor, reporter:
   }
 
-  private val uncaughtExceptionHandler: Thread.UncaughtExceptionHandler = new Thread.UncaughtExceptionHandler {
+  private[this] val uncaughtExceptionHandler: Thread.UncaughtExceptionHandler = new Thread.UncaughtExceptionHandler {
```
nice catch

adriaanm replied:
@paulp's, actually

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -675,11 +675,115 @@ object Future {
       throw new IllegalStateException("problem in scala.concurrent internal callback", t)
+
+    /**
+     * The BatchingExecutor trait had to be inlined into InternalCallbackExecutor for binary compatibility.
```
hmmm, what happened to the test case that came with this bugfix?

adriaanm replied:
test/files/jvm/scala-concurrent-tck.scala? it's still there

viktorklang replied:
Ah, just making sure :-) Them tests are nice things I hear!

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -675,11 +675,115 @@ object Future {
+      }
+    }
+
+    // protected def unbatchedExecute(r: Runnable): Unit
```
delete

adriaanm commented:
re: `scala.reflect.runtime`

This is not exactly a name that lowers expectations of binary compatibility. We already have `scala.reflect.internal` for that. Anything not in a package called "internal" has to be compatible.

adriaanm committed:
[nomaster] inline importPrivateWithinFromJavaFlags into SymbolTable

This reworks 02ed5fb so that we don't change JavaUniverse's super classes.

This is necessary to maintain binary compatibility with 2.10.0.

adriaanm commented:
that said, it's unfortunate classes in scala.runtime inherit from classes in scala.internal
this is problematic as illustrated in c53b9c5

xeno-by commented:
@adriaanm reflect.internal and reflect.runtime being internal was the status quo since the last summer (e.g. take a look at the scaladoc). That's what we also discussed on the mailing list a week or two ago. Why change the policy all of a sudden a day before the release?

xeno-by commented:
1) Oh man, I was totally wrong yesterday. `scala.reflect.runtime.universe` is of type `api.JavaUniverse`, so it doesn't leak the underlying `runtime.JavaUniverse` to the outer world: https://github.com/scalamacros/kepler/blob/9164c2af4183e04986eb652e09081daba3d0d279/src/reflect/scala/reflect/runtime/package.scala#L16.

2) See the Scaladoc for `scala.reflect.runtime`. The only visible parts are the package class, which exposes the public API, and `JavaUniverse`, which explicitly says "Should not be instantiated directly". The rest in that package is `private[scala]`, `private[reflect]` and `private[runtime]`.

Therefore I would argue that everything except `scala.reflect.runtime.package` should be considered internal.

adriaanm commented:
I wasn't aware that the runtime package was considered internal. It's
pretty confusing to have it part internal, part public, though. Also note
that those private[package] modifiers you mention are not relevant for
binary compatibility. They are not expressible in byte code.

On Sunday, February 10, 2013, Eugene Burmako wrote:

> 1) Oh man, I was totally wrong yesterday. scala.reflect.runtime.universeis of type
> api.JavaUniverse, so it doesn't leak the underlying runtime.JavaUniverseto the outer world:
> https://github.com/scalamacros/kepler/blob/9164c2af4183e04986eb652e09081daba3d0d279/src/reflect/scala/reflect/runtime/package.scala#L16
> .
> 
> 2) See the Scaladoc for scala.reflect.runtime. The only visible parts are
> the package class, which exposes the public API, and JavaUniverse, which
> explicitly says "Should not be instantiated directly". The rest in that
> package is private[scala], private[reflect] and private[runtime].
> 
> Therefore I would argue that everything except
> scala.reflect.runtime.package should be considered internal.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2103#issuecomment-13345891..

xeno-by commented:
How would you suggest to improve the situation? We could, for example, move all the stuff except `package.scala` to `scala.reflect.runtime.internal`.

adriaanm commented:
The goal for minor releases of 2.10 is for them to be drop-in replacements.
Are you comfortable with changing things in the reflect.runtime
package, guaranteeing  that
nothing will break?

On Sunday, February 10, 2013, Adriaan Moors wrote:

> I wasn't aware that the runtime package was considered internal. It's
> pretty confusing to have it part internal, part public, though. Also note
> that those private[package] modifiers you mention are not relevant for
> binary compatibility. They are not expressible in byte code.
> 
> On Sunday, February 10, 2013, Eugene Burmako wrote:
> 
> > 1) Oh man, I was totally wrong yesterday. scala.reflect.runtime.universeis of type
> > api.JavaUniverse, so it doesn't leak the underlying runtime.JavaUniverseto the outer world:
> > https://github.com/scalamacros/kepler/blob/9164c2af4183e04986eb652e09081daba3d0d279/src/reflect/scala/reflect/runtime/package.scala#L16
> > .
> > 
> > 2) See the Scaladoc for scala.reflect.runtime. The only visible parts
> > are the package class, which exposes the public API, and JavaUniverse,
> > which explicitly says "Should not be instantiated directly". The rest in
> > that package is private[scala], private[reflect] and private[runtime].
> > 
> > Therefore I would argue that everything except
> > scala.reflect.runtime.package should be considered internal.
> > 
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2103#issuecomment-13345891..

xeno-by commented:
Absolutely. Those are just implementation classes, not used anywhere except for `scala.reflect.runtime.package`.

adriaanm commented:
It's late here, unfortunately. We'll discuss more tomorrow.

On Sunday, February 10, 2013, Adriaan Moors wrote:

> The goal for minor releases of 2.10 is for them to be drop-in
> replacements. Are you comfortable with changing things in the
> reflect.runtime package, guaranteeing  that nothing will break?
> 
> On Sunday, February 10, 2013, Adriaan Moors wrote:
> 
> > I wasn't aware that the runtime package was considered internal. It's
> > pretty confusing to have it part internal, part public, though. Also note
> > that those private[package] modifiers you mention are not relevant for
> > binary compatibility. They are not expressible in byte code.
> > 
> > On Sunday, February 10, 2013, Eugene Burmako wrote:
> > 
> > > 1) Oh man, I was totally wrong yesterday. scala.reflect.runtime.universeis of type
> > > api.JavaUniverse, so it doesn't leak the underlying runtime.JavaUniverseto the outer world:
> > > https://github.com/scalamacros/kepler/blob/9164c2af4183e04986eb652e09081daba3d0d279/src/reflect/scala/reflect/runtime/package.scala#L16
> > > .
> > > 
> > > 2) See the Scaladoc for scala.reflect.runtime. The only visible parts
> > > are the package class, which exposes the public API, and JavaUniverse,
> > > which explicitly says "Should not be instantiated directly". The rest in
> > > that package is private[scala], private[reflect] and private[runtime].
> > > 
> > > Therefore I would argue that everything except
> > > scala.reflect.runtime.package should be considered internal.
> > > 
> > > —
> > > Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2103#issuecomment-13345891..

xeno-by commented:
I see. Thanks for your time and good night!

adriaanm commented:
I see the docs say "In Scala 2.10.0, the Scala Reflection API and its implementation have an "experimental" status. This means that the API and the docs are not complete and can be changed in binary- and source-incompatible manner in 2.10.1. "

This is not what I thought we had decided, when we said that by transitivity even code that doesn't directly use reflection will end up requiring scala-reflect, so that it would have to provide the same kind of compatibility as the rest of the standard library. I'm willing to make an exception for `scala.reflect.internal`, as changes there should only cause breakage when users mix and match classes in scala.reflect.\* from different Scala versions.

Binary compatibility errors affect everybody, so I'm not willing to compromise it (more than allowing changes in scala.reflect.internal), simply to save under 1% of our fixed bugs.

New feature development goes into 2.11.x. 2.10.x is our stable version now. Seriously.

retronym commented:
That decision was confirmed after the 2.10.0 release.

I also don't think these particular bugs are worth delaying the release for. But changing goalposts are not conducive  to productivity, we must improve at pinning them down earlier.

adriaanm commented:
I don't want to be lugging around goal posts. Especially not last minute. You're right, that's not awesome.

Though I agree we didn't make this explicit enough and I have again failed to communicate it properly,
I thought the thread https://groups.google.com/forum/#!topic/scala-internals/IApLHqVt-9M was a pretty clear indicator that we were going to be serious about compatibility _everywhere_. I see scala.reflect.internal as a valid compromise, but the package that holds the very entry point to reflection should not break across minor version bumps. Especially, as you said, not for the sake of one or two bugs out of almost 200 fixed since 2.10.0.

adriaanm commented:
Greg has worked last week on adding filtering to MiMa so that it can again be part of pr validation as soon as his PR is merged and MiMa 0.15 is released. I should have made this a priority sooner, so that we wouldn't have to do this now.

xeno-by commented:
So it doesn't matter than compat violations in `scala.reflect.runtime` don't touch public classes? What difference does it make whether these classes are located next to the entry point into reflection or not?

Also I'd like to note that with this policy in place, we most likely won't be able to make reflection thread-safe in 2.10.x, as doing that would almost certainly entail changing the implementation of runtime reflection.

Finally it looks like I failed to document the assumptions that were always in the air (and which I'm positive we discussed when the whole forward compat thing started brewing). Now when I try to search for scala.reflect.runtime and compatibility I only get http://grokbase.com/p/gg/scala-internals/13221vzfzd/binary-compat, which is a week old.

retronym commented:
Let's take up the discussion at the Scala meeting on Tuesday with a view towards 2.10.2.

gkossakowski commented:
@retronym: I'm afraid that single meeting might not be enough for that. I wanted to start discussion long time ago about this here: https://groups.google.com/d/msg/scala-internals/9x3EYhbwFvI/TlgvZvI-e-EJ but I dropped the ball due to all the other things I had to do at the time.

I think the biggest problem is with aggregating all the relevant data. I'll try to do that before Tuesday but I think during the meeting all we should do is to just probe more people and get more data but _not_ make any decisions yet. By next week we might have enough of material (also input from community) to make final call.

Also, I agree that we should target 2.10.2 with any decisions. On 2.10.1 we just need to be conservative.

retronym commented:
One point for the discussion: I don't think that the "inliner" argument applies to this package as the code is only exposed via an abstract interface. At stake is the ability to bring thread safety to 2.10.2 reflection, so we must explore this fully and sensibly.

adriaanm commented:
Yes. As always this is about trade offs. As others have said, let's be
conservative in 2.10.1. A massive amount of work has gone into all these
bug fixes. Let's ensure adaptors of 2.10.0 will have a great experience
upgrading to 2.10.1. 2.10.2 is not far away. I'm willing to discuss moving
it one month closer if we feel that alleviates the high pressure we put on
every point release. I don't think cramming as much as we can in 2.10.1 is
a good way to approach this.

On Sunday, February 10, 2013, Jason Zaugg wrote:

> One point for the discussion: I don't think that the "inliner" argument
> applies to this package as the code is only exposed via an abstract
> interface. At stake is the ability to bring thread safety to 2.10.2
> reflection, so we must explore this fully and sensibly.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2103#issuecomment-13360995..
